test-utils/src/mollusk.rs

```diff
@@ -119,64 +130,61 @@ pub fn mollusk_add_so_files(
     });
 }
 
-/// Returns `(accounts before, exec result)`
+/// On success:
+/// - returns `Ok(ExecOk)` with resulting accounts as `AccountMap`
+/// - asserts lamports are balanced
+/// - asserts all resulting accounts are rent-exempt
+///
+/// On failure:
+/// - returns `Err(ProgramResult)` with the failure
 pub fn mollusk_exec(
     svm: &Mollusk,
-    ix: &Instruction,
-    onchain_state: &AccountMap,
-) -> (AccountMap, InstructionResult) {
-    let mut keys: Vec<_> = ix.accounts.iter().map(|a| a.pubkey).collect();
-    keys.sort_unstable();
-    keys.dedup();
-
-    let accs_bef: AccountMap = keys
+    ixs: &[Instruction],
+    accs_bef: &AccountMap,
+) -> Result<ExecOk, ProgramResult> {
+    let mut keys: Vec<_> = ixs
         .iter()
-        .map(|k| {
-            let (k, v) = onchain_state.get_key_value(k).unwrap();
-            (*k, v.clone())
-        })
+        .flat_map(|ix| ix.accounts.iter().map(|a| a.pubkey))
         .collect();
-
-    let accs_vec: Vec<_> = accs_bef.iter().map(|(k, v)| (*k, v.clone())).collect();
-
-    let res = svm.process_instruction(ix, &accs_vec);
-
-    (accs_bef, res)
-}
-
-/// Like `mollusk_exec` but with validation checks applied to resulting accounts.
-/// Returns `(accounts before, exec result)`.
-pub fn mollusk_exec_validate(
-    svm: &Mollusk,
-    ix: &Instruction,
-    onchain_state: &AccountMap,
-    checks: &[Check],
-) -> (AccountMap, InstructionResult) {
-    let mut keys: Vec<_> = ix.accounts.iter().map(|a| a.pubkey).collect();
     keys.sort_unstable();
     keys.dedup();
 
-    let accs_bef: AccountMap = keys
+    let accs_vec: Vec<_> = keys
         .iter()
         .map(|k| {
-            let (k, v) = onchain_state.get_key_value(k).unwrap();
+            let (k, v) = accs_bef.get_key_value(k).unwrap();
             (*k, v.clone())
         })
         .collect();
 
-    let mut accs_vec: Vec<_> = accs_bef.iter().map(|(k, v)| (*k, v.clone())).collect();
-    accs_vec.sort_by_key(|(k, _)| *k);
+    let res = svm.process_instruction_chain(ixs, &accs_vec);
 
-    let res = svm.process_and_validate_instruction(ix, &accs_vec, checks);
+    if res.program_result.is_ok() {
+        let resulting_accounts: AccountMap = res.resulting_accounts.iter().cloned().collect();
 
-    (accs_bef, res)
+        assert_balanced(accs_bef, &resulting_accounts);
+        assert!(
+            res.run_checks(&[Check::all_rent_exempt()], &svm.config, svm),
+            "Not all accounts are rent-exempt after execution"
+        );
+
+        Ok(ExecOk {
+            resulting_accounts,
+            compute_units_consumed: res.compute_units_consumed,
+            execution_time: res.execution_time,
+            raw_result: res.raw_result,
+            return_data: res.return_data,
+        })
+    } else {
+        Err(res.program_result)
```

Comment:
i wrote `ProgramError | InstructionError`, not `ProgramResult` for a reason. `ProgramResult::Success` shouldnt be returned if execution failed